File: solutions/144-binary-tree-preorder-traversal/binary-tree-preorder-traversal.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function preorderTraversal(root: TreeNode | null): number[] {

    const res: number[] = [];

    function helper(node: TreeNode | null): void {
        if(node === null) {
            return;
        }

        res.push(node.val);
        helper(node.left);
        helper(node.right);
    }

    helper(root);
    
    return res;
    
};

File: solutions/144-binary-tree-preorder-traversal/binary-tree-preorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def preorderTraversal(self, root: Optional[TreeNode]) -> List[int]:

        result = []
        self.helper(root, result)
        return result
        
    def helper(self, node: Optional[TreeNode], vals: List[int]):
        if node is None:
            return
        vals.append(node.val)
        self.helper(node.left, vals)
        self.helper(node.right, vals)    